app/models/space.rb

```diff
@@ -4,7 +4,28 @@ class Space < ApplicationRecord
   validates :name, presence: true
   belongs_to :parent, inverse_of: :children, foreign_key: :space_id, class_name: 'Space', optional: true
   has_many :children, inverse_of: :parent, foreign_key: :space_id, class_name: 'Space'
-  has_many :contracts
   has_many :assets
+  has_many :contract_spaces
```

Comment:
`has_and_belongs_to_many :contracts`

http://guides.rubyonrails.org/association_basics.html#the-has-many-association
2.6 The has_and_belongs_to_many Association
A has_and_belongs_to_many association creates a direct many-to-many connection with another model, with no intervening model. For example, if your application includes assemblies and parts, with each assembly having many parts and each part appearing in many assemblies, you could declare the models this way: